File: src/main/java/catwalks/block/BlockScaffolding.java
package catwalks.block;

import catwalks.Conf;
import catwalks.Const;
import catwalks.EnumCatwalkMaterial;
import catwalks.register.ItemRegister;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

public class BlockScaffolding extends BlockBase {
		
	public final int page;
	
	public BlockScaffolding(int page) {
		super(Material.IRON, "scaffold" + page);
		this.page = page;
		setHardness(0.1f);
	}
	
	@Override
	public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World world, BlockPos pos, EntityPlayer player) {
		return new ItemStack(ItemRegister.scaffold, 1, state.getValue(Const.MATERIAL).ordinal());
	}
	
	@Override
	public float getPlayerRelativeBlockHardness(IBlockState state, EntityPlayer playerIn, World worldIn, BlockPos pos) {
		
		if(playerIn.inventory.getCurrentItem() != null){
			Item item = playerIn.inventory.getCurrentItem().getItem();
			
			if( item == Item.getItemFromBlock(this) && playerIn.isSneaking()) {
				return super.getPlayerRelativeBlockHardness(state, playerIn, worldIn, pos);
			}
			
			if( item == ItemRegister.tool) {
				return 1;
			}
		}
//		if(state.getValue(Const.MATERIAL_META) == EnumCatwalkMaterial.WOOD) {
//			return 1;
//		}
		return super.getPlayerRelativeBlockHardness(state, playerIn, worldIn, pos);
	}
	
	public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
    {
		if(this.harvesters.get() == null) {
			super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
			return;
		}
		
        if (!worldIn.isRemote && !worldIn.restoringBlockSnapshots) // do not drop items while restoring blockstates, prevents item dupe
        {
            java.util.List<ItemStack> items = getDrops(worldIn, pos, state, fortune);
            chance = net.minecraftforge.event.ForgeEventFactory.fireBlockHarvesting(items, worldIn, pos, state, fortune, chance, false, harvesters.get());

            for (ItemStack item : items)
            {
                if (worldIn.rand.nextFloat() <= chance)
                {
                    spawnAsEntityNoPickupDelayExactPos(worldIn, this.harvesters.get().getPositionVector().addVector(0, 0.25, 0), item);
                }
            }
        }
    }
	
	public static void spawnAsEntityNoPickupDelayExactPos(World worldIn, Vec3d pos, ItemStack stack)
    {
        if (!worldIn.isRemote && worldIn.getGameRules().getBoolean("doTileDrops") && !worldIn.restoringBlockSnapshots) // do not drop items while restoring blockstates, prevents item dupe
        {
            if (captureDrops.get())
            {
                capturedDrops.get().add(stack);
                return;
            }
            EntityItem entityitem = new EntityItem(worldIn, pos.xCoord, pos.yCoord, pos.zCoord, stack);
            entityitem.setPickupDelay(0);
            worldIn.spawnEntityInWorld(entityitem);
        }
    }
	
	@Override
	public int damageDropped(IBlockState state) {
		return getMetaFromState(state);
	}
	
	@Override
	public boolean canHarvestBlock(IBlockAccess world, BlockPos pos, EntityPlayer player) {
		return true;
	}
	
	{ /* state stuff */ }
	
	@Override
	protected BlockStateContainer createBlockState() {
		return new BlockStateContainer(this, Const.MATERIAL);
	}

	@Override
	public IBlockState getStateFromMeta(int meta) {
		return getDefaultState().withProperty(Const.MATERIAL, EnumCatwalkMaterial.values()[page << 4 & meta]);
	}
	
	@Override
	public int getMetaFromState(IBlockState state) {
		return state.getValue(Const.MATERIAL).ordinal() & 0b1111;
	}
	
	{ /* rendering stuff */ }
	
	@Override
	public BlockRenderLayer getBlockLayer() {
		return BlockRenderLayer.CUTOUT;
	}
	
	@Override
	public boolean isOpaqueCube(IBlockState state) {
		return false;
	}
	
	@Override
	public boolean isFullCube(IBlockState state) {
		return false;
	}
	
	@Override
	@SideOnly(Side.CLIENT)
    public boolean shouldSideBeRendered(IBlockState state, IBlockAccess worldIn, BlockPos pos, EnumFacing side)
    {
		if(Conf.showScaffoldInsideFaces)
			return super.shouldSideBeRendered(state, worldIn, pos, side);
        return worldIn.getBlockState(pos.offset(side)).getBlock() == this ? false : super.shouldSideBeRendered(state, worldIn, pos, side);
    }

	{ /* ladder */ }
	
//	@Override
//	public boolean shouldApplyClimbing(World world, BlockPos pos, EntityLivingBase entity) {
//		return ( entity.moveForward != 0 || entity.moveStrafing != 0 );// && world.getBlockState(pos).getValue(Const.MATERIAL_META) == EnumCatwalkMaterial.WOOD;
//	}
//
//	@Override
//	public boolean shouldApplyFalling(World world, BlockPos pos, EntityLivingBase entity) {
//		return true;//world.getBlockState(pos).getValue(Const.MATERIAL_META) == EnumCatwalkMaterial.WOOD;
//	}
//
//	@Override
//	public double climbSpeed(World world, BlockPos pos, EntityLivingBase entity) {
//		return 1;
//	}
//
//	@Override
//	public double fallSpeed(World world, BlockPos pos, EntityLivingBase entity) {
//		return 1;
//	}
//
//	@Override
//	public double horizontalSpeed(World world, BlockPos pos, EntityLivingBase entity) {
//		return Double.POSITIVE_INFINITY;
//	}
}

File: src/main/java/catwalks/block/BlockScaffolding.kt
package catwalks.block

import catwalks.Const
import catwalks.EnumCatwalkMaterial
import catwalks.part.PartScaffold
import catwalks.part.data.ScaffoldRenderData
import catwalks.register.ItemRegister
import catwalks.util.GeneralUtil
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.BlockRenderLayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.RayTraceResult
import net.minecraft.util.math.Vec3d
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.common.property.ExtendedBlockState
import net.minecraftforge.common.property.IExtendedBlockState

class BlockScaffolding(val page: Int) : BlockBase(Material.IRON, "scaffold" + page) {

    init {
        setHardness(0.1f)
    }

    override fun getPickBlock(state: IBlockState, target: RayTraceResult?, world: World, pos: BlockPos, player: EntityPlayer?): ItemStack {
        return ItemStack(ItemRegister.scaffold, 1, state.getValue(Const.MATERIAL).ordinal)
    }

    override fun damageDropped(state: IBlockState): Int {
        return state.getValue(Const.MATERIAL).ordinal
    }

    override fun createBlockState(): BlockStateContainer {
        return ExtendedBlockState(this, arrayOf(Const.MATERIAL), arrayOf(Const.SCAFFOLD_RENDER_DATA))
    }

    override fun getExtendedState(state: IBlockState, world: IBlockAccess, pos: BlockPos): IBlockState {
        if(state !is IExtendedBlockState)
            return state

        val arr = BooleanArray(EnumFacing.values().size) { true }

        for(dir in EnumFacing.values()) {
            val poff = pos.offset(dir)
            if(world.isSideSolid(poff, dir.opposite, false)) {
                arr[dir.ordinal] = false
            }
            if(arr[dir.ordinal]) {
                val adjacent = world.getBlockState(poff)
                val block = adjacent.block
                if(block is BlockScaffolding) {
                    if(adjacent.getValue(Const.MATERIAL) == state.getValue(Const.MATERIAL)) {
                        arr[dir.ordinal] = false
                    }
                }
            }
            if(arr[dir.ordinal]) {
                val part = GeneralUtil.getPart(PartScaffold::class.java, world, poff)
                if(part != null && part.catwalkMaterial == state.getValue(Const.MATERIAL)) {
                    arr[dir.ordinal] = false
                }
            }
        }

        return state.withProperty(Const.SCAFFOLD_RENDER_DATA, ScaffoldRenderData(arr))
    }

    override fun getStateFromMeta(meta: Int): IBlockState {
        var i = (page shl 4) or meta
        if(i >= EnumCatwalkMaterial.values().size)
            i = 0
        var mat = EnumCatwalkMaterial.values()[i]
        if(mat !in Const.MATERIAL.allowedValues)
            mat = EnumCatwalkMaterial.CUSTOM_0
        return defaultState.withProperty(Const.MATERIAL, mat)
    }

    override fun getMetaFromState(state: IBlockState): Int {
        return state.getValue(Const.MATERIAL).ordinal and 15
    }

    override fun getBlockLayer(): BlockRenderLayer {
        return BlockRenderLayer.CUTOUT
    }

    override fun isOpaqueCube(state: IBlockState?): Boolean {
        return false
    }

    override fun isFullCube(state: IBlockState?): Boolean {
        return false
    }

    companion object {

        fun spawnAsEntityNoPickupDelayExactPos(worldIn: World, pos: Vec3d, stack: ItemStack) {
            if (!worldIn.isRemote && worldIn.gameRules.getBoolean("doTileDrops") && !worldIn.restoringBlockSnapshots)
            // do not drop items while restoring blockstates, prevents item dupe
            {
                if (Block.captureDrops.get()) {
                    Block.capturedDrops.get().add(stack)
                    return
                }
                val entityitem = EntityItem(worldIn, pos.xCoord, pos.yCoord, pos.zCoord, stack)
                entityitem.setPickupDelay(0)
                worldIn.spawnEntityInWorld(entityitem)
            }
        }
    }

    //	@Override
    //	public boolean shouldApplyClimbing(World world, BlockPos pos, EntityLivingBase entity) {
    //		return ( entity.moveForward != 0 || entity.moveStrafing != 0 );// && world.getBlockState(pos).getValue(Const.MATERIAL_META) == EnumCatwalkMaterial.WOOD;
    //	}
    //
    //	@Override
    //	public boolean shouldApplyFalling(World world, BlockPos pos, EntityLivingBase entity) {
    //		return true;//world.getBlockState(pos).getValue(Const.MATERIAL_META) == EnumCatwalkMaterial.WOOD;
    //	}
    //
    //	@Override
    //	public double climbSpeed(World world, BlockPos pos, EntityLivingBase entity) {
    //		return 1;
    //	}
    //
    //	@Override
    //	public double fallSpeed(World world, BlockPos pos, EntityLivingBase entity) {
    //		return 1;
    //	}
    //
    //	@Override
    //	public double horizontalSpeed(World world, BlockPos pos, EntityLivingBase entity) {
    //		return Double.POSITIVE_INFINITY;
    //	}
}
